landlab/components/profiler/channel_profiler.py

```diff
@@ -0,0 +1,462 @@
+# coding: utf8
+# ! /usr/env/python
+"""channel_profiler.py component to create channel profiles."""
+
+import numpy as np
+
+from landlab import RasterModelGrid
+from landlab.components.profiler.base_profiler import _BaseProfiler
+
+
+class ChannelProfiler(_BaseProfiler):
+    """Extract and plot the long profiles within a drainage network.
+
+    The ChannelProfiler can work on grids that have used route-to-one or
+    route-to-multiple flow directing. This component expects that all of the
+    following at-node grid fields will be present:
+
+        'drainage_area'
+        'flow__receiver_node'
+        'flow__link_to_receiver_node'
+
+    These are typically created by using the FlowAccumulator component.
+
+    Four parameters control the behavior of this component. By default, the
+    component will extract the single largest channel in the drainage network
+    up to a drainage area below a threshold of two times the cell area.
+
+    Providing a value for the parameter ``threshold`` changes this value.
+    Setting ``main_channel_only`` to ``False`` will plot all of the channels in
+    each watershed that have drainage area below the threshold. The number of
+    watersheds can be changed from one based on the parameter
+    ``number_of_watersheds``. By default the ChannelProfiler will identify the
+    largest watersheds and use them. If instead you want to extract the channel
+    network draining to specific nodes, you can provide values for the
+    parameter ``starting_nodes``. The length of ``starting_nodes`` must be
+    consistent with the value of ``number_of_watersheds``.  The node IDs of the
+    ``starting_nodes`` do not need to be on the domain boundaries or the
+    outlets of watersheds. They do need to have drainage area above the
+    indicated threshold.
+
+    The ``run_one_step`` method extracts the channel network and stores it in
+    the ``profile_structure`` and ``distances_upstream`` bound properties. To
```

Comment:
@SiccarPoint A couple of thoughts about your questions on the profile structure, segments etc. 

No matter the answer, the documentation will be improved, but the extent to which depends a bit on the answer. 

Do you think the profile structure is something that is user facing? I had originally made it user facing, but upon thinking more, I am starting to think that it shouldn't be. 

Why does a user need to know the specifics of the data structure that holds channel segment ids and distances upstream? I don't think they do. A developer who wants to use this to construct a Network Model Grid does, but not a user? Correct me if you think otherwise.

Reply:
I think it should be. Plenty of cases where the user would want to plot, e.g., S-A along the stream line, so needs access to the node IDs; `profile_structure[0][0]` if a single stream line, but needing to iterate on `profile_structure[0]` if in segments. If they don't know the structure, it's going to be tough to know how to work with what pops out.

I think just cleaning and expanding the documentation with reference to segments will clean up a lot of this!

Reply:
Thanks for the feedback. I think I'm going to change it to a dictionary with the following structure:

```python
profiler.channel_structure = {
    outlet1_id : {
        (segment1_downstream_id, segment1_upstream_id): 
            {"ids": [list of all nodes in segment 1],
             "distances": [list of distances upstream],
             "color": (optional, default is same for each outlet)
            },
        (segment2_downstream_id, segment2_upstream_id): same datastructure as above,
         },
    outlet2_id : {
        (segment1_downstream_id, segment1_upstream_id): same datastructure as above,
        (segment2_downstream_id, segment2_upstream_id): same datastructure as above,
    }
}

My thinking is that there is no real ordering with the segments within a single watershed (to answer your prior question). Which makes it hard to document and hard to use. But the "true" ordering has to do with the connectivity between the segment end nodes (so I use it as a tuple for the dictionary key). This then makes things 
- easier to use to make a network model grid
- easier to document (since there is a more systematic structure)
- keeps the upstream, distances, and color information all together (rather than having multiple parallel data structures)

What do you think?